### ISSUE TITLE
Mongo OOMs during sync for large databases

### ISSUE THREAD

**Issue by sbelak** [metabase/metabase#11983]
On large DBs (say 15k "tables") Mongo OOMs during sync which results in some fields being partially synced or not at all.

Stack trace is inconclusive, but for instance all fields get fingerprinted as having distinct count = 0.

MB  v1.34.2

**Comment by flamber** [comment]
This could be because sync currently gets all documents on sync - instead of just the first 10k, which are the only ones used:
https://github.com/metabase/metabase/blob/ce939219f8a3e28ecb00deb232154780a1dcfcef/modules/drivers/mongo/src/metabase/driver/mongo.clj#L180-L198

Looking at [Mongo profiler](https://stackoverflow.com/questions/15204341/mongodb-logging-all-queries), it looks like it's initially getting 101 documents:
<details>

```
{
    "op" : "query",
    "ns" : "sample.orders",
    "command" : {
        "find" : "orders",
        "$db" : "sample",
        "$readPreference" : {
            "mode" : "primaryPreferred"
        }
    },
    "cursorid" : 22655095939.0,
    "keysExamined" : 0,
    "docsExamined" : 101,
    "numYield" : 0,
    "nreturned" : 101,
    "locks" : {
        "Global" : {
            "acquireCount" : {
                "r" : NumberLong(3)
            }
        },
        "Database" : {
            "acquireCount" : {
                "r" : NumberLong(1)
            }
        },
        "Collection" : {
            "acquireCount" : {
                "r" : NumberLong(1)
            }
        }
    },
    "storage" : {},
    "responseLength" : 16018,
    "protocol" : "op_msg",
    "millis" : 0,
    "planSummary" : "COLLSCAN",
    "execStats" : {
        "stage" : "COLLSCAN",
        "nReturned" : 101,
        "executionTimeMillisEstimate" : 0,
        "works" : 102,
        "advanced" : 101,
        "needTime" : 1,
        "needYield" : 0,
        "saveState" : 1,
        "restoreState" : 0,
        "isEOF" : 0,
        "invalidates" : 0,
        "direction" : "forward",
        "docsExamined" : 101
    },
    "ts" : ISODate("2021-05-21T09:42:48.341Z"),
    "client" : "172.17.0.1",
    "allUsers" : [ 
        {
            "user" : "metabase",
            "db" : "admin"
        }
    ],
    "user" : "metabase@admin"
}
```

</details>
But then a following getmore-request asks for all documents:
<details>

```
{
    "op" : "getmore",
    "ns" : "sample.orders",
    "command" : {
        "getMore" : NumberLong(22655095939),
        "collection" : "orders",
        "$db" : "sample",
        "$readPreference" : {
            "mode" : "primaryPreferred"
        }
    },
    "originatingCommand" : {
        "find" : "orders",
        "$db" : "sample",
        "$readPreference" : {
            "mode" : "primaryPreferred"
        }
    },
    "cursorid" : 22655095939.0,
    "keysExamined" : 0,
    "docsExamined" : 18659,
    "cursorExhausted" : true,
    "numYield" : 145,
    "nreturned" : 18659,
    "locks" : {
        "Global" : {
            "acquireCount" : {
                "r" : NumberLong(148)
            }
        },
        "Database" : {
            "acquireCount" : {
                "r" : NumberLong(146)
            }
        },
        "Collection" : {
            "acquireCount" : {
                "r" : NumberLong(146)
            }
        }
    },
    "storage" : {},
    "responseLength" : 2971313,
    "protocol" : "op_msg",
    "millis" : 8,
    "planSummary" : "COLLSCAN",
    "execStats" : {
        "stage" : "COLLSCAN",
        "nReturned" : 18760,
        "executionTimeMillisEstimate" : 0,
        "works" : 18762,
        "advanced" : 18760,
        "needTime" : 1,
        "needYield" : 0,
        "saveState" : 146,
        "restoreState" : 146,
        "isEOF" : 1,
        "invalidates" : 0,
        "direction" : "forward",
        "docsExamined" : 18760
    },
    "ts" : ISODate("2021-05-21T09:42:48.352Z"),
    "client" : "172.17.0.1",
    "allUsers" : [ 
        {
            "user" : "metabase",
            "db" : "admin"
        }
    ],
    "user" : "metabase@admin"
}
```

</details>

**Comment by luizarakaki** [comment]
@camsaul Do you think we can close this with https://github.com/metabase/metabase/issues/27045?

This looks hard repro and now we are reading 1k docs instead of 10k per collection.

**Comment by camsaul** [comment]
Not hard to repro at all. But @metamben fixed this in #27137, the query we use now includes an explicit limit.

Fixed by #27137
